cores/esp8266/umm_malloc/umm_malloc.cpp

```diff
@@ -1600,6 +1731,29 @@ static void *_umm_realloc( void *ptr, size_t size ) {
 
     c = umm_assimilate_down(c, 0);
 
+    /*
+     * Currently all or most of the heap has been grabbed. Do an early split of
+     * allocation down to the amount needed to do a successful move operation.
+     * This will allow an alloc/new from a ISR to succeed while a memmove is
+     * running.
+     */
+    if( (UMM_NBLOCK(c) - c) > prelimBlockSize ) {
+        umm_make_new_block( c, prelimBlockSize, 0, 0 );
+        _umm_free( (void *)&UMM_DATA(c+prelimBlockSize) );
```

Comment:
Isn't this call to umm free() inside the umm realloc critical section? That would mean there is still nesting present.

Reply:
Yes, I didn't like that either; however, if we unlock outside the critical section, and we take an interrupt. Then, that ISR could try and malloc memory and get an OOM result because at this moment we have claimed all the free memory above and below us. Which might be most of the memory that is left, depending on fragmentation. So I had to move the `free` just before the critical section exit.

And of course, the improved critical section supports nesting.

Reply:
I don't think it's necessary to move the critical section locks to remove the nesting, but rather refactor _umm_free into two functions: a clre function that does no locking or arg checking, and a wrapper function that checks the args and locks before calling the core and unlocks afterwards. Then, this call here would be replaced with the core call, i. e. no locking and hence no nested critical sections.